contracts/distribution/PushDistributor.sol

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.20;
+
+import { Party } from "./../party/Party.sol";
+import { PartyGovernance } from "./../party/PartyGovernance.sol";
+import { ProposalExecutionEngine } from "./../proposals/ProposalExecutionEngine.sol";
+import { IERC20 } from "../tokens/IERC20.sol";
+import { LibERC20Compat } from "./../utils/LibERC20Compat.sol";
+import { ReentrancyGuard } from "openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract PushDistributor is ReentrancyGuard {
+    event Distributed(Party party, IERC20 token, address[] members, uint256 amount);
+
+    error NotEnoughETH(uint256 expectedAmount, uint256 receivedAmount);
+    error WrongProposalId(uint256 proposalId);
+    error WrongMembers();
+    error MembersNotSorted();
+
+    using LibERC20Compat for IERC20;
+
+    // Token address used to indicate ETH.
+    IERC20 private constant ETH_ADDRESS = IERC20(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE);
+
+    function distribute(
```

Comment:
Would like detailed natspec here. It's a pretty advanced function. Otherwise PR looks good.